### PR TITLE
Enable state for roles in func2protocol decorator

### PR DESCRIPTION
**Pull request by Jasopaum**
## Description

Add a states arg in func2protocol decorator.
We may want to think about how this State will be used.

## Type of change

Please mark the options that are relevant.

- [ ] Added/Modified tutorials
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

* [x] I have added tests for my changes
* [x] I did follow the [contribution guidelines](https://github.com/OpenMined/PySyft/blob/master/CONTRIBUTING.md)
* [x] I have commented my code following [Google style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

### CODE REVIEW

File: test/execution/test_protocol.py [hunk 1]
```diff
@@ -231,6 +231,28 @@ def protocol(roles):
...
 
+def test_stateful_protocol(workers):
+    shapes = {"alice": ((1,),), "bob": ((1,),)}
+    states = {"alice": (th.tensor([1]), th.tensor([3])), "bob": (th.tensor([5]),)}
+
+    @sy.func2protocol(args_shape=shapes, states=states)
+    def protocol(roles):
+        # fetch tensors from stores
+        # TODO will we fetch tensors that come from states this way?
+        # With tags?
```
Comment (karlhigley) [inline test/execution/test_protocol.py:242]: 🤔 

For `Plans`, they're fetched with `state.read()`, right? I remember you saying you weren't a fan of that, so maybe this is a good time to think about how to do it better.

Comment (Jasopaum) [reply test/execution/test_protocol.py:242]: Yep, they're fetched with `read`. As we are going towards something very similar for other tensors fetched from the worker store, I think we'll need to talk about differences between them/if we can totally merge these 2 concepts (from an API point of view at least).

Comment (karlhigley) [reply test/execution/test_protocol.py:242]: Maybe in the Syft Core sync tomorrow?

**Comment by karlhigley** [comment]
I suppose we could call the method that reads from `State` objects something like `load_state()`. 🤔

**Comment by Jasopaum** [comment]
> I suppose we could call the method that reads from `State` objects something like `load_state()`. 🤔

You mean replacing `state.read()` or `role.load()`?

**Comment by karlhigley** [comment]
> You mean replacing `state.read()` or `role.load()`?

I was thinking we'd end up with `role.load()` loading from an `ObjectStore` and `role.load_state()` loading from `State`. Not sure if that makes any sense, but it might?

**Comment by Jasopaum** [comment]
> > You mean replacing `state.read()` or `role.load()`?
> 
> I was thinking we'd end up with `role.load()` loading from an `ObjectStore` and `role.load_state()` loading from `State`. Not sure if that makes any sense, but it might?

It might! But I still don't know how these 2 are related...
The wires behind the `role.load()` aren't connected yet. I can implement a first `load_state` though, searching by tags, what do you think?

**Comment by karlhigley** [comment]
Sounds okay to me! Do we even need tags yet though? 🤷

**Comment by Jasopaum** [comment]
Oh, you want to write `a, b, c = role.load_state()`?
Yeah, it's definitely possible!

**Comment by karlhigley** [comment]
I mean, I'm not sure that _want_ to, but it would be fine for now as an incremental step. 😄

**Pull request merged**